### PR TITLE
Sign in | All tests

### PR DESCRIPTION
**Pull request by vitali-akulau**
- Autotests implementation from Sign In suite
---
client 2021-05-14 15-30-56

### CODE REVIEW

File: client/cypress/.eslintrc.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
   },
   "rules": {
     "import/no-named-as-default": 0,
-    "import/no-named-as-default-member": 0
-}
+    "import/no-named-as-default-member": 0,
+    "no-use-before-define": ["error", { "functions": false, "classes": true }]
```
Comment (s34rching) [inline client/cypress/.eslintrc.json:11]: nice

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 2]
```diff
@@ -1,18 +1,169 @@
...
+      .should('be.visible')
+      .get(SignInPage.signUpPasswordField)
+      .should('be.visible')
+      .get(SignInPage.signUpConfirmPasswordField)
+      .should('be.visible');
+  }
+
+  function assertUserLogged() {
+    cy.url().should('include', HOME_PAGE_URL_PATH)
+      .get(HomePage.signOutLink)
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js]: I believe this is the only thing which is sufficient to tell that user has been logged in

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 3]
```diff
@@ -1,18 +1,169 @@
...
       .should('be.visible')
       .and('contain.text', SIGN_OUT_LINK_TEXT);
-  });
+  }
+
+  function assertUserLoggedOut() {
+    cy.url().should('eq', HOME_PAGE_URL_PATH)
+      .get(HomePage.signInHeaderLink)
+      .should('be.visible')
+      .and('contain.text', SIGN_IN_LINK_TEXT);
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js]: there are too much assertions
please do not forget that something might `not` be on the page
make it more efficient

File: client/cypress/model/SignInPage.js [hunk 4]
```diff
@@ -11,14 +20,28 @@ class SignInPage {
...
+    return this;
+  }
+
   insertField(element, value){
     cy.get(element)
-      .clear()
-      .type(value);
+      .clear().then(e => {
+        if (value !== '')
+          cy.wrap(e).type(value)
```
Comment (s34rching) [inline client/cypress/model/SignInPage.js]: what is the reason of wrapping?

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 5]
```diff
@@ -1,18 +1,169 @@
...
+  function assertUserIsNotAuthorized() {
+    cy.url().should('include', '/signing')
+      .get(SignInPage.signUpNameField)
+      .should('be.visible')
+      .get(SignInPage.signUpEmailField)
+      .should('be.visible')
+      .get(SignInPage.signUpPasswordField)
+      .should('be.visible')
+      .get(SignInPage.signUpConfirmPasswordField)
+      .should('be.visible');
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js]: oh man
you should not chain methods this way b/c you are disabling Cypress waiting for conditions
we need to hop on the call to let me explain it to you

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 6]
```diff
@@ -1,18 +1,169 @@
...
+  });
+
+  function assertUserNotLoggedIn() {
+    cy.url().should('include', '/signing')
+      .get(SignInPage.signInEmailField)
+      .should('be.visible')
+      .get(SignInPage.signInPasswordField)
+      .should('be.visible')
+      .get(SignInPage.signInButton)
+      .should('be.visible');
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js]: oh man
you should not chain methods this way b/c you are disabling Cypress waiting for conditions
we need to hop on the call to let me explain it to you

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 7]
```diff
@@ -1,18 +1,169 @@
...
-import validUser from '../../fixtures/users';
 import HomePage from '../../model/HomePage';
+import user from '../../fixtures/user';
 
 const SIGN_OUT_LINK_TEXT = 'SIGN OUT';
+const SIGN_IN_LINK_TEXT = 'SIGN IN';
+const WARNING_TEXT = 'The email address is already in use by another account.';
+const HOME_PAGE_URL_PATH = 'http://127.0.0.1:3000/';
 
 describe('Sign in page', () => {
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js:5]: please try to wrap tests into their separate `describe` blocks to distinct `sign-in` from `sign-up` tests

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 8]
```diff
@@ -1,18 +1,169 @@
 import SignInPage from '../../model/SignInPage';
-import validUser from '../../fixtures/users';
 import HomePage from '../../model/HomePage';
+import user from '../../fixtures/user';
 
 const SIGN_OUT_LINK_TEXT = 'SIGN OUT';
+const SIGN_IN_LINK_TEXT = 'SIGN IN';
+const WARNING_TEXT = 'The email address is already in use by another account.';
+const HOME_PAGE_URL_PATH = 'http://127.0.0.1:3000/';
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js]: create another helper file which contains all these hard-coded things

File: client/cypress/fixtures/user.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  email: 'test@test.com',
+  newEmail: `test.user@test${Date.now()}.com`,
+  wrongEmail: 'email.com',
+  emptyEmail: '',
+  password: 123456,
+  wrongPassword: '-',
+  emptyPassword: '',
+  confirmPassword: 123456,
+  emptyConfirmPassword: '',
+  differentConfirmPassword: 12345,
```
Comment (s34rching) [inline client/cypress/fixtures/user.js]: you cannot iterate over your fixture the way you've composed values
my suggestion is to create array containing various `user` objects
then you can provide the data and run them using any of iteration methods

File: client/cypress/model/SignInPage.js [hunk 10]
```diff
@@ -11,14 +20,28 @@ class SignInPage {
     return this;
   }
 
+  fillSignUpForm(name, email, password, confirmPassword){
+    this.insertField(this.signUpNameField, name);
+    this.insertField(this.signUpEmailField, email);
+    this.insertField(this.signUpPasswordField, password);
+    this.insertField(this.signUpConfirmPasswordField, confirmPassword);
+    return this;
```
Comment (s34rching) [inline client/cypress/model/SignInPage.js]: what is the sacred sense of returning `this`?

**Review by s34rching — CHANGES_REQUESTED** [review]
I need to explain few concepts of Cypress that you should always remember about

File: client/cypress/integration/authentication/sign-in.spec.js (hunk 2 above)
Comment (vitali-akulau) [reply client/cypress/integration/authentication/sign-in.spec.js]: Agree. Have fixed

File: client/cypress/model/SignInPage.js (hunk 4 above)
Comment (vitali-akulau) [reply client/cypress/model/SignInPage.js]: I use test data from user.js. Some keys have empty value. If I try pass empty value in the tests I get an exception.
client 2021-05-14 16-48-07
The current implementation makes it possible not to pass empty values

File: client/cypress/model/SignInPage.js (hunk 10 above)
Comment (vitali-akulau) [reply client/cypress/model/SignInPage.js]: Using 'this' get a possibility create a chain calls for one instance only. 
For example:
With 'this'
react-app – sign-in spec js 2021-05-14 17-02-50
Without 'this'
react-app – sign-in spec js 2021-05-14 17-04-44
Using this, you can chain calls for a single instance, even if the test flow goes through multiple pages.
Technically, it doesn't do anything. But with this approach, you see beautifully designed code

File: client/cypress/fixtures/helper.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+const helper = {
```
Comment (s34rching) [inline client/cypress/fixtures/helper.js]: `constants`?

**Review by s34rching — COMMENTED** [review]
Just comment

File: client/cypress/fixtures/constants.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+const constants = {
```
Comment (ivankatsenya) [inline client/cypress/fixtures/constants.js]: looks better

File: client/src/components/sign-in/sign-in.component.jsx [hunk 13]
```diff
@@ -27,7 +27,7 @@ export const SignIn = ({ googleSignInStart, emailSignInStart }) => {
 
   return (
     <SignInContainer>
-      <SignInTitle>I already have an account</SignInTitle>
+      <SignInTitle data-test="sign-in-title">I already have an account</SignInTitle>
```
Comment (ivankatsenya) [inline client/src/components/sign-in/sign-in.component.jsx:30]: but do not abuse attributes adding regarding there might not be any or adding attributes will be blocked for you

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 14]
```diff
@@ -1,18 +1,122 @@
...
+  }
+
+  function assertUserIsNotAuthorized() {
+    cy.url().should('include', '/signing');
+    cy.get(SignInPage.signUpTitle).should('be.visible');
+  }
+
+  function assertUserLogged() {
+    cy.url().should('include', constants.homePageUrlPath);
+  }
```
Comment (ivankatsenya) [inline client/cypress/integration/authentication/sign-in.spec.js]: this condition is always truthy, b/c you are checking that url contains baseUrl :)
you should check UI to assert user is logged in

File: client/cypress/fixtures/constants.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+  signInLinkText: 'SIGN IN',
+  warningText: 'The email address is already in use by another account.',
+  homePageUrlPath: 'http://127.0.0.1:3000/',
```
Comment (ivankatsenya) [inline client/cypress/fixtures/constants.js]: this is `baseUrl` - add it to the `cypress.json`

Comment (ivankatsenya) [reply client/cypress/fixtures/constants.js]: the only thing is that you should not add `/` from url

**Review by ivankatsenya — CHANGES_REQUESTED** [review]
it needs more updates

File: client/cypress/integration/authentication/sign-in.spec.js [hunk 16]
```diff
@@ -1,18 +1,121 @@
...
+    cy.url().should('include', '/signing');
+    cy.get(SignInPage.signInTitle).should('be.visible');
+  }
+
+  function assertUserIsNotAuthorized() {
+    cy.url().should('include', '/signing');
+    cy.get(SignInPage.signUpTitle).should('be.visible');
+  }
+
+  function assertUserLogged() {
```
Comment (s34rching) [inline client/cypress/integration/authentication/sign-in.spec.js:110]: `assertUserLoggedIn`

**Review by s34rching — APPROVED** [review]
Okay, good to go for now!

**Pull request merged**